Add optional UV2 logic for lightmapping to primitive shapes

BastiaanOlij opened the pull request:
This PR introduces optionally adding UV2 texture coordinates to primitive shapes including support for padding along the seams of the mesh. It will also set the `lightmap size hint` based on a texel size that can be configured in project settings:

![image](https://user-images.githubusercontent.com/1945449/199367239-a3f71c07-1f49-4f60-a053-c1ce1060ff1b.png)

The default values of this is `0.2`.

The Primitive Meshes related to particles have not been changed. 

This PR supersedes #49932

BastiaanOlij committed:
Add optional UV2 logic for lightmapping to primitive shapes

jcostello commented:
For what I can recall, emissive materials were working on baked lightmaps

Calinou commented:
> For what I can recall, emissive materials were working on baked lightmaps

Emissive materials in Godot 4 require the mesh with the emissive material to have UV2, and have its GI mode set to **Static**. In Godot 3, meshes that don't have UV2 but had **Use In Baked Light** enabled could still *contribute* to lightmaps.

BastiaanOlij commented:
Before this gets merged, had some more thoughts about improving the UV algorithm.
The current calculation for UV1 is based on texture maps that have even sides and should remain as is for backwards compatibility.
But it would be easy to adjust the calculations for UV2 to have the UVs proportional to the actual size of the faces which is much better suited for lightmapping or texture painting. So I want to make that change before we merge.

Also on chat there was a suggestion to automatically assign the `lightmap size hint` to something appropriate, we need to decide what this calculation is (basically we need to design on a pixel per meter metric).

Calinou commented:
> Also on chat there was a suggestion to automatically assign the lightmap size hint to something appropriate, we need to decide what this calculation is (basically we need to design on a pixel per meter metric).

I wonder how XAtlas determines this. It uses a texels per meter metric that is defined in the 3D scene's import options. Since PrimitiveMesh doesn't have import options, we can add such a property to PrimitiveMesh instead.

BastiaanOlij commented:
@Calinou hmm, I wonder if it doesn't make more sense to make it a project setting, else we'll start getting questions of people wanting it to calculate backwards when they update the lightmap size and we'll need to do some sort of aspect ratio fixation.

BastiaanOlij commented:
Ok adjusted most UV2 calculations so they are based on actual sizes of the objects, just need to do `BoxMesh` and `PrismMesh`.

Also still need to add `lightmap size hint` calculation.

@Calinou the texelsize mentioned, if it's 0.2 does that mean that 1 pixel is 0.2 x 0.2 sized? So a square meter would have a 5x5 pixel grid for lighting data?

Calinou commented:
> @Calinou the texelsize mentioned, if it's 0.2 does that mean that 1 pixel is 0.2 x 0.2 sized? So a square meter would have a 5x5 pixel grid for lighting data?

It means each texel square is 0.2×0.2 units. Larger values result in *less* detailed lightmapping, as the object will have fewer texels on the lightmap texture.

BastiaanOlij commented:
I've changed this back to draft as I put the final things to it. It's been a bit of puzzling. I'm changing the UV2 calculation to be based on actual dimensions of the object and to update the lightmap size hint based on a new project setting.

So far I've adjusted:
- BoxMesh
- CylinderMesh
- CapsuleMesh

![image](https://user-images.githubusercontent.com/1945449/199248174-8a6b7524-6c70-414f-9677-40c9b6582734.png)

BastiaanOlij commented:
Finished prism and torus, I don't think the meshes for particles/trails require the UV2 treatment so I've left those to the default logic.

Prism and Box meshes could use some improvement to better pack the faces and prioritize sizes for smaller textures but I'll leave that for someone with more time on their hands :P

clayjohn commented on scene/resources/primitive_meshes.h:
```diff
@@ -152,8 +167,10 @@ class BoxMesh : public PrimitiveMesh {
+
 public:
-	static void create_mesh_array(Array &p_arr, Vector3 size, int subdivide_w = 0, int subdivide_h = 0, int subdivide_d = 0);
+	static void create_mesh_array(Array &p_arr, Vector3 size, int subdivide_w = 0, int subdivide_h = 0, int subdivide_d = 0, bool p_add_uv2 = false, const float p_uv2_padding = 0.0);
```
Should this default to a padding value of ``1.0`` like the others?

BastiaanOlij replied:
Not sure, the default is to not add padding so the function doesn't change its current behavior. But I guess setting it to 1.0 in case just `p_add_uv2` is set to true...

BastiaanOlij replied:
ah, I really just used the 0.0 default in one place now did I... weird :) I'll make it consistent :)

clayjohn commented on scene/resources/primitive_meshes.cpp:
```diff
@@ -94,6 +97,26 @@ void PrimitiveMesh::_update() const {
 
+	if (add_uv2) {
+		// _create_mesh_array should populate our UV2, this is a fallback in case it doesn't.
+		// As we don't know anything about the geometry we only bad the right and bottom edge
```
```suggestion
		// As we don't know anything about the geometry we only pad the right and bottom edge
```

clayjohn reviewed (approved):
Implementation looks fine to me, we need @reduz to approve the API additions though

BastiaanOlij commented:
Android build failure seems to be an issue with the CI, not the code, @akien-mga ?

akien-mga commented:
Thanks!
